crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,477 @@
+    ) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u8::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.insert(utxo_id, node_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.insert(*contract_id, node_id);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    /// Clear the caches of the storage when a transaction is removed.
+    fn clear_cache(&mut self, outputs: &[Output], tx_id: &TxId) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u16::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.remove(&utxo_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.remove(contract_id);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    fn get_inner(&self, index: &NodeIndex) -> Result<&StorageData, Error> {
+        self.graph
+            .node_weight(*index)
+            .ok_or(Error::TransactionNotFound(format!(
+                "Transaction with index {:?} not found",
+                index
+            )))
+    }
+
+    fn get_dependents_inner(&self, index: &NodeIndex) -> Result<Vec<NodeIndex>, Error> {
+        Ok(self
+            .graph
+            .neighbors_directed(*index, petgraph::Direction::Outgoing)
+            .collect())
+    }
+}
+
+impl Storage for GraphStorage {
+    type StorageIndex = NodeIndex;
+
+    fn store_transaction(
+        &mut self,
+        transaction: PoolTransaction,
+        dependencies: Vec<Self::StorageIndex>,
+        collided_transactions: Vec<Self::StorageIndex>,
+    ) -> Result<(Self::StorageIndex, RemovedTransactions), Error> {
+        let tx_id = transaction.id();
+
+        // Remove collisions and their dependencies from the graph
+        let mut removed_transactions = vec![];
+        for collision in collided_transactions {
+            removed_transactions
+                .extend(self.remove_node_and_dependent_sub_graph(collision)?);
+        }
```

Comment:
Moved it after the checks

Reply:
Moving this logic after the check means that:
If the current subtree is full, but removing the collided transaction will free space, we will not do that. It is why I asked what behavior do we want to support?=)


We also can try to predict what size of substree will be and do this check before as well. But maybe it is to complicated

Reply:
I think we want to support this behavior as it will prevent a new one to be inserted with a correct error returned rather than having your existing transaction being deleted in the future for no reason (because we ended up not adding the new one).

Of course simulating would be the best and I might do it when I will work on the new insertion rules with new limits

Reply:
Second thought : When removing collision it will not free space in the dependency chain because is it deletes any dependency then the TX can't be included anymore so the collision deletion doesn't impact the tx_chain_length